x-pack/plugins/lens/public/editor_frame_plugin/state_management.ts

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { EditorFrameProps } from './editor_frame';
+
+export interface EditorFrameState {
+  activeDatasource: string | null;
+  datasourceState: unknown;
+  datasourceIsLoading: boolean;
+
+  activeVisualization: string | null;
+  visualizationState: {
+    [visualizationId: string]: unknown;
+  };
+}
+
+export type Action =
+  | {
+      type: 'UPDATE_DATASOURCE_STATE';
+      newState: unknown;
+    }
+  | {
+      type: 'UPDATE_VISUALIZATION_STATE';
+      newState: unknown;
+    }
+  | {
+      type: 'SWITCH_VISUALIZATION';
+      newVisulizationId: string;
+      initialState: unknown;
+    }
+  | {
+      type: 'SWITCH_DATASOURCE';
+      newDatasourceId: string;
+    };
+
+export const getInitialState = (props: EditorFrameProps) => {
+  return {
+    datasourceState: null,
```

Comment:
We basically store multiple visualization states to be able to switch back and forth between them without loosing any configuration settings which are not related to the table columns.
Example: You have an xy chart with stack bars and switch to a pie chart. Now you switch back to the xy chart, but the settings of using stacked bars has been deleted and you are back to unstacked lines.

We do this for the visualizations because it's really common to switch around between them, but it's not common to switch between datasources after you started configuring. In fact this might not even really make sense, because they are just so different - what should happen when we switch from index pattern datasource to csv datasource? 

Working with two top level props looks nice though, I changed the shape to
```
export interface EditorFrameState {
  visualization: {
    activeId: string | null;
    stateMap: {
      [visualizationId: string]: unknown;
    };
  };
  datasource: {
    activeId: string | null;
    state: unknown;
    isLoading: boolean;
  };
}
```

I also renamed all props which contain a map of things (`Record<string, T>`) to ...Map to make the relationship clear.